pkg/controller/clusterstate/clusterstate_controller_test.go

```diff
@@ -0,0 +1,246 @@
+package clusterstate
+
+import (
+	"context"
+	"reflect"
+	"sort"
+	"testing"
+
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/hive/pkg/apis"
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+)
+
+const (
+	testName                 = "cluster1"
+	testNamespace            = "cluster1namespace"
+	testKubeconfigSecretName = "kubeconfig-secret"
+)
+
+func TestClusterStateReconcile(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+	configv1.Install(scheme.Scheme)
+
+	log.SetLevel(log.DebugLevel)
+
+	cs := func(t *testing.T, c client.Client) *hivev1.ClusterState {
+		st := &hivev1.ClusterState{}
+		err := c.Get(context.TODO(), types.NamespacedName{Namespace: testNamespace, Name: testName}, st)
+		if err != nil {
+			t.Fatalf("unexpected: %v", err)
+		}
+		return st
+	}
+	co := clusterOperator
+	uco := unavailableClusterOperator
+
+	tests := []struct {
+		name     string
+		existing []runtime.Object
+		remote   []runtime.Object
+		validate func(*testing.T, client.Client, reconcile.Result)
+		noUpdate bool
+	}{
+		{
+			name: "update empty clusterstate",
+			existing: []runtime.Object{
+				testClusterState(),
+				testClusterDeployment(),
+				testKubeconfigSecret(),
+			},
+			remote: []runtime.Object{co("a"), co("b"), co("c")},
+			validate: func(t *testing.T, c client.Client, result reconcile.Result) {
+				st := cs(t, c)
+				validateStatus(t, st.Status, co("a"), co("b"), co("c"))
+			},
+		},
+		{
+			name: "steady state",
+			existing: []runtime.Object{
+				testClusterStateWithStatus(co("d"), co("e")),
+				testClusterDeployment(),
+				testKubeconfigSecret(),
+			},
+			remote: []runtime.Object{co("d"), co("e")},
+			validate: func(t *testing.T, c client.Client, result reconcile.Result) {
+				st := cs(t, c)
+				validateStatus(t, st.Status, co("d"), co("e"))
+			},
+			noUpdate: true,
+		},
+		{
+			name: "changed state",
+			existing: []runtime.Object{
+				testClusterStateWithStatus(co("a"), co("b"), co("c")),
+				testClusterDeployment(),
+				testKubeconfigSecret(),
+			},
+			remote: []runtime.Object{co("a"), co("b"), uco("c")},
+			validate: func(t *testing.T, c client.Client, result reconcile.Result) {
+				st := cs(t, c)
+				validateStatus(t, st.Status, co("a"), co("b"), uco("c"))
+			},
+		},
```

Comment:
Add unit tests for
1) removed remote clusteroperator,
2) added remote condition, and
3) removed remote condition.